teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,431 @@
+   param name prefixes to group related params (e.g. `ssh-`, `okra-vm`, `okra-token`).
+
+## Requirements
+
+* Must be possible to programmatically determine the structure of the object
+* Must be possible for a result object to be empty
+* Must be possible to use the object results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has a object param
+* Must be possible to use one specific value in a object result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value for the specified key, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+
+## Proposal
+
+* We would add support for object types for results and params, in addition to the existing string and array support
+  * Initially we would only support string keys, eventually we can expand this to all values (string, array, object)
+    (note that this is the case for arrays as well, we don't yet support arrays of arrays)
+* We would use [json object schema syntax](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526)
+  to express the object structure
+* To support object results, we would support writing results to a `.json` file, as described in
+  [TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477)
+
+This feature would be considered alpha and would be
+[gated by the alpha flag](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).
+
+### Notes/Caveats
+
+[See also TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for notes and caveats
+specific to supporting json results.
+
+* What if a Pipeline tries to use a object key that doesn't exist in a param or result?
+  * Some of this could be caught at creation time, i.e. pipeline params would declare the keys each object param will
+    contain, and we could validate at creation time that those keys exist
+  * For invalid uses that can only be caught at runtime, the PipelineRun would fail
+* What if a Task tries to write more keys to a result than the result declares it contains?
+  * The TaskRun would fail (alternatively we could decide to ignore the extra keys)
+* What if a Task tries to write less keys to a result than the result declares it contains?
+  * The TaskRun would fail
+
+### Risks and Mitigations
+
+[See also TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for risks and mitigations
+specific to expanding ArrayOrString to support more types.
+
+### User Experience
+
+* Using the json schema syntax to describe the structure of the object is a bit verbose and might not be intuitive
+  to users; see [Alternatives](#alternatives) for discussion of an option where we create our own syntax.
```

Comment:
What I was thinking is sort of inline with the idea of SchemeRefs, but built-in because we need to know the type/scheme to support substitution.  This feels like a nice way to test the waters with schema refs for fairly well defined things.

Reply:
Pursuing schema refs and having known schema types sounds like a great next step after this TEP - I'll include it in a "future work" section in the TEP. Having well known types like `GitRepo` that ship with Tekton would be really useful (esp for tools built on Tekton, like chains). If these types are built into Tekton I think the next question is what to do when people want to provide their own (also a problem we had with PipelineResources - adding your own PipelineResources required changing the Tekton Controller) - maybe people can make their own `ParamSchema` instances - and/or maybe we could use something as simple as a ConfigMap?

I think @mattmoor 's idea above goes a bit further than this and suggests being able to use these schemas to define the body of the Tasks themselves, e.g reusable steps which sounds to me like another proposal entirely (which maybe again could leverage this one?) and reminds me of [TEP-0054 step reuse](https://github.com/tektoncd/community/pull/369) - and it also sounds like we're starting to talk about going beyond just "variable replacement" and into the realms of templating 🤔